### PR TITLE
refactor(expr): separate function implementations into a new crate

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Fresh build times (starting from risingwave_pb) before and after this PR:

build-cmp

This PR separates function implementations from the expr crate into a new `expr_impl` crate. The purpose is to parallize the build of functions and subsequent modules, reduce overall build time and speedup debugging cycle.

In details, we move as many functions as possible to the new crate, and leave only interfaces and a few basic expressions, such as InputRef, Literal and UDF in the original crate. To make all functions available, one has to depend on the `expr_impl` crate and call `risingwave_expr_impl::enable!()` to ensure that the library is linked into the final binary. We have added it to the batch, stream and frontend crate as `#[dev-dependencies]` to let unit tests work, and added it to the cmd, cmd_all, sqlsmith and simulation crate to let final artifacts work.

Other minor changes:
- rename module `agg` to `aggregate`, and `vector_op` to `scalar`.
- rename data types in function signatures `int16/int32/int64/float32/float64` to `int2/int4/int8/float4/float8`.
- remove `str_to_date/time/timestamp` functions and reimplement them as `FromStr for Date/Time/Timestamp`. 

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

### CODE REVIEW

File: src/common/src/types/datetime.rs [hunk 1]
```diff
@@ -75,31 +70,169 @@ impl_chrono_wrapper!(Date, NaiveDate);
...
-enum InvalidParamsErrorKind {
+enum ErrorKind {
     #[error("Invalid date: days: {days}")]
     Date { days: i32 },
     #[error("Invalid time: secs: {secs}, nanoseconds: {nsecs}")]
     Time { secs: u32, nsecs: u32 },
     #[error("Invalid datetime: seconds: {secs}, nanoseconds: {nsecs}")]
     DateTime { secs: i64, nsecs: u32 },
+    #[error("Can't cast string to date (expected format is YYYY-MM-DD)")]
+    ParseDate,
```
Comment (BugenZhao) [inline src/common/src/types/datetime.rs:214]: Reminds me of https://github.com/risingwavelabs/risingwave/pull/11420 🥹

File: src/expr/impl/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+mod aggregate;
+mod scalar;
+mod table_function;
+
+/// Enable functions in this crate.
+#[macro_export]
+macro_rules! enable {
+    () => {
+        #[allow(unused_extern_crates)]
+        extern crate risingwave_expr_impl;
```
Comment (BugenZhao) [inline src/expr/impl/src/lib.rs:44]: What about `use risingwave_expr_impl as _`?

**Review by BugenZhao — COMMENTED** [review]
Looks exciting!

**Comment by xxchan** [comment]
(Off topic) This is scary. In https://github.com/risingwavelabs/risingwave/issues/9553, expr was 14s and connector was 12s. Now they are 41 and 35s? 🥶

**Review by xxchan — COMMENTED** [review]
> Other minor changes:

I really don't like minor changes included in a large and important PR. 😕😕😕 (especially since GitHub review UX sucks) For me, the efforts of reviewing PRs scales *quadratically* with the number of ~~lines~~ units of changes.

File: src/expr/core/src/expr/build.rs [hunk 3]
```diff
@@ -126,7 +126,8 @@ pub(super) fn get_children_and_return_type(prost: &ExprNode) -> Result<(&[ExprNo
 /// Build an expression from a string.
 ///
 /// # Example
-/// ```
+///
+/// ```ignore
```
Comment (BugenZhao) [inline src/expr/core/src/expr/build.rs:130]: Why do we ignore this doc test?

**Review by BugenZhao — APPROVED** [review]
LGTM!

**Review by stdrc — APPROVED** [review]
LSTM!

**Comment by wangrunji0408** [comment]
> (Off topic) This is scary. In #9553, expr was 14s and connector was 12s. Now they are 41 and 35s? 🥶

I'm not sure if it is the codegen approach in `#[function]` macro that increases the compilation time. 🥵
Anyway, as we are adding more and more functions and connectors, their build time will increase inevitably. What we can do is not to let them block the entire building process and to parallelize them as much as possible.

> I really don't like minor changes included in a large and important PR. 😕😕😕 (especially since GitHub review UX sucks) For me, the efforts of reviewing PRs scales quadratically with the number of lines units of changes.

Sorry for the trouble I made. I did this way because I usually mix these changes in a branch, but am too lazy to split them up and then resolve conflicts to merge them back. I'll try my best to provide a better review experience in the future. 🤡

**Comment by xxchan** [comment]
> I'm not sure if it is the codegen approach in #[function] macro that increases the compilation time. 🥵

Tried to profile it, but didn't come up with any useful conclusion yet. 🤔️

flamegrame from the data by
```
cargo rustc -p risingwave_expr_impl -- -Z self-profile
```


rustc

```
cargo llvm-lines -p risingwave_expr_impl
```

```
  Lines                  Copies               Function name
  -----                  ------               -------------
  3056366                69242                (TOTAL)
   188370 (6.2%,  6.2%)    971 (1.4%,  1.4%)  <core::slice::iter::Iter<T> as core::iter::traits::iterator::Iterator>::fold
   164787 (5.4%, 11.6%)   4374 (6.3%,  7.7%)  core::option::Option<T>::map
   152475 (5.0%, 16.5%)   1194 (1.7%,  9.4%)  alloc::vec::Vec<T,A>::extend_trusted
   147059 (4.8%, 21.4%)   1301 (1.9%, 11.3%)  <alloc::vec::Vec<T> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<T,I>>::from_iter
   118300 (3.9%, 25.2%)    910 (1.3%, 12.6%)  risingwave_expr::expr::Expression::eval_v2::{{closure}}
   107926 (3.5%, 28.8%)   4364 (6.3%, 18.9%)  alloc::boxed::Box<T>::new
    60861 (2.0%, 30.7%)   1271 (1.8%, 20.8%)  core::iter::adapters::map::map_fold::{{closure}}
    44002 (1.4%, 32.2%)   1194 (1.7%, 22.5%)  alloc::vec::Vec<T,A>::extend_trusted::{{closure}}
    43914 (1.4%, 33.6%)   1271 (1.8%, 24.3%)  <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
    39120 (1.3%, 34.9%)   3260 (4.7%, 29.0%)  alloc::boxed::Box<T>::pin
    35217 (1.2%, 36.1%)   1215 (1.8%, 30.8%)  core::ops::function::FnOnce::call_once
    34211 (1.1%, 37.2%)   1201 (1.7%, 32.5%)  core::iter::traits::iterator::Iterator::for_each
    28273 (0.9%, 38.1%)    358 (0.5%, 33.0%)  <core::iter::adapters::enumerate::Enumerate<I> as core::iter::traits::iterator::Iterator>::next
    24737 (0.8%, 38.9%)   1364 (2.0%, 35.0%)  core::iter::traits::iterator::Iterator::map
    24337 (0.8%, 39.7%)   1201 (1.7%, 36.8%)  core::iter::traits::iterator::Iterator::for_each::call::{{closure}}
    23080 (0.8%, 40.5%)     40 (0.1%, 36.8%)  <core::iter::adapters::flatten::FlattenCompat<I,U> as core::iter::traits::iterator::Iterator>::size_hint
    21581 (0.7%, 41.2%)    211 (0.3%, 37.1%)  <core::iter::adapters::zip::Zip<A,B> as core::iter::adapters::zip::ZipImpl<A,B>>::next
    16033 (0.5%, 41.7%)    232 (0.3%, 37.5%)  core::iter::traits::iterator::Iterator::fold
    14560 (0.5%, 42.2%)    910 (1.3%, 38.8%)  risingwave_expr::expr::Expression::eval_v2
    13610 (0.4%, 42.6%)    179 (0.3%, 39.0%)  risingwave_common::array::bool_array::BoolArray::from_iter_bitmap
....
```

**Comment by wangrunji0408** [comment]
The macro generates unique code for each function. There's no template instantiation, so it won't show in llvm-lines reports.  🤔

**Comment by xxchan** [comment]
It takes ~6s to expand macros

image

**Comment by xxchan** [comment]
Well, macro expand time seems to be caused by `#[async_trait]` 🤡

**Comment by TennyZhuang** [comment]
> Well, macro expand time seems to be caused mainly by `#[async_trait]` 🤡

Due to the dyn usage, we can’t remove the macro here.

**Comment by BugenZhao** [comment]
> > Well, macro expand time seems to be caused mainly by `#[async_trait]` 🤡
> 
> Due to the dyn usage, we can’t remove the macro here.

That's true. However, I'm wondering if we can try expanding the async trait for implementors ourselves to speed up compile time, as we already utilize codegen for functions.

**Comment by xxchan** [comment]
BTW, I just tried things like
```
[profile.dev.package.async-trait]
opt-level = 3
[profile.dev.package.ctor]
opt-level = 3
[profile.dev.package.syn]
opt-level = 3
[profile.dev.package.quote]
opt-level = 3
[profile.dev.package.proc-macro2]
opt-level = 3
[profile.dev.package.risingwave_expr_macro]
opt-level = 3
[profile.dev.package.risingwave_expr]
opt-level = 3
```
This doesn't help much. 🤔️

**Pull request merged**